msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/stress/AbstractMsalUiStressTest.java

```diff
@@ -0,0 +1,264 @@
+import com.microsoft.identity.internal.testutils.labutils.LabUserQuery;
+
+import org.junit.Before;
+import org.junit.rules.RuleChain;
+import org.junit.rules.Timeout;
+
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+
+public abstract class AbstractMsalUiStressTest<T, S> extends AbstractMsalUiTest {
+
+    // Sets the interval duration in seconds to which the device performance will be monitored.
+    private static final long DEVICE_MONITOR_INTERVAL = TimeUnit.SECONDS.toMillis(1);
+    private static final String DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSZ";
+    private Exception executionException;
+
+    private int testsPassed = 0;
+    private int testsFailed = 0;
+
+    @Override
+    @Before
+    public void setup() {
+        super.setup();
+        executionException = null;
+
+        DeviceMonitor.setProfiler(PerformanceProfile.CPU, new CPUMonitor());
+        DeviceMonitor.setProfiler(PerformanceProfile.MEMORY, new MemoryMonitor());
+        DeviceMonitor.setProfiler(PerformanceProfile.NETWORK, new NetworkUsageMonitor());
+    }
+
+    /**
+     * Run the stress tests
+     */
+    public void run() throws Exception {
+        final T prerequisites = this.prepare();
+
+        final BlockingQueue<Runnable> blockingQueue = new ArrayBlockingQueue<>(getNumberOfThreads());
+        final RejectedExecutionHandler rejectedExecutionHandler = new ThreadPoolExecutor.CallerRunsPolicy();
+
+        final ExecutorService executorService = new ThreadPoolExecutor(
+                1,
+                getNumberOfThreads() + 1,
+                0L,
+                TimeUnit.MILLISECONDS,
+                blockingQueue,
+                rejectedExecutionHandler
+        );
+
+
+        final long startTime = System.currentTimeMillis();
+        final long timeLimit = TimeUnit.MINUTES.toMillis(getTimeLimit());
+
+        // use the execution thread pool to collect device stats
+        executorService.submit(getPerformanceStatsCollector());
+
+
+        while (System.currentTimeMillis() - startTime < timeLimit && executionException == null) {
+            executorService.submit(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        S result = execute(prerequisites);
+
+                        boolean passed = isTestPassed(result);
+
+                        updateTestPassRate(passed);
+                    } catch (Exception ex) {
+                        executionException = ex;
+                    }
+                }
+            });
+            TimeUnit.MILLISECONDS.sleep(50);
+        }
+
+        executorService.shutdown();
```

Comment:
The service can actually fail to shut down.  We should check the return, and then call shutdownNow if it does fail, and fail ourselves.